### DISCUSSION TITLE
How to run pyaedt script on an already loaded project.

### DISCUSSION THREAD

**Discussion by amichel0205** [ansys/pyaedt#4112] [Show and tell]
# Question: How to run pyaedt script on an already loaded project.

Answer: I can see 3 ways to do it:
1. Using the console
2. Run a python script:
      - Using PyAEDT Menu select "Run PyAEDT script"
      - Adding an icon linked to a specific script

## Using the console

- AEDT is launched and a project is loaded
- In PyAEDT Menu select "Console"

![image](https://github.com/ansys/pyaedt/assets/119294121/3a3421e3-aa63-407d-9108-e5c1aeb069f1)

- Once the Console is open you can type the following
  - import pyaedt
  - from pyaedt import Hfss3dLayout*
  -  h3dl = Hfss3dLayout()
- The command Hfss3dLayout() will instantiate the active project and active design
- Enter PyAEDT command like you would do in a script

![image](https://github.com/ansys/pyaedt/assets/119294121/18576d88-ce05-4d1c-9432-e0dc5558043a)

- *Use the relevant import:
 -Hfss for HFSS
 -Hfss3dLayout for HFSS 3D Layout
 -…

- Now you can enter any command like in a script
- Below you see a sample of pyaedt command in the console
  - List of designs from the project
  - Design name of the active design
  - Display design name in the logger
  - Add setup
  - Add  sweep to the setup

![image](https://github.com/ansys/pyaedt/assets/119294121/6f458474-30d3-4e82-ac39-58ea0302f9fd)

-When you want to stop using the console you have to type:
  -**h3dl.save_project()** if you want to save
  -**h3d.release_desktop(close_projects=False, close_desktop=False)** if you want to keep the project open

> [!CAUTION]
> **It is mandatory to release the desktop** so AEDT can be released. 
>If you don't do it, you will not be able to close AEDT after stopping the console (except from killing the process from the task manager). >In case you open the console and don't do anything you will have to enter the following command:
>
>desktop.release_desktop(close_projects=False, close_on_exit=False)
>
>For more information on command, you can look in documentation: https://aedt.docs.pyansys.com/version/stable/

Project after closing the console

![image](https://github.com/ansys/pyaedt/assets/119294121/631a989a-95fb-4d9f-bc36-184103ee5829)

### Using PyAEDT Menu select "Run PyAEDT script

-AEDT is launched and a project is loaded
-In PyAEDT Menu select "Run PyAEDT script"

![image](https://github.com/ansys/pyaedt/assets/119294121/d93fed11-c7bb-48ae-8fe8-d6c6c1352086)

-From the window select the script to run and click Open

![image](https://github.com/ansys/pyaedt/assets/119294121/41b36a82-7315-4d47-9f82-780311594b5e)

Below is a simple script that will create and add a setup in HFSS 3D Layout design:

```python
import pyaedt

h3dl = pyaedt.Hfss3dLayout()

h3dl.logger.info(h3dl.design_name)
setup = h3dl.create_setup()
h3dl.create_linear_count_sweep(
    setupname=setup.name,
    unit="GHz",
    freqstart=3,
    freqstop=7,
    num_of_freq_points=1001,
    sweepname="sweep1",
    sweep_type="Interpolating",
    interpolation_tol_percent=1,
    interpolation_max_solutions=255,
    save_fields=False,
    use_q3d_for_dc=False,
)

h3dl.save_project()
h3dl.release_desktop(close_projects=False, close_desktop=False)
```

Project after script is executed

![image](https://github.com/ansys/pyaedt/assets/119294121/9374ceab-5978-4f8d-8c85-fe8b74597c4f)

### Adding a Icon linked to a specific script

-It is possible to add an icon in the “Automation” bar. This is documented here:
https://aedt.docs.pyansys.com/version/0.7/API/_autosummary/pyaedt.desktop.Desktop.add_script_to_menu.html#pyaedt.desktop.Desktop.add_script_to_menu

- AEDT is launched and a project is loaded
- In PyAEDT Menu select "Console" and add the command **desktop.add_script_to_menu**
- Here an icon “AddSetup” will be added to the Automation bar. 
- Script **test_interactive.py** is the same script used in the previous section.

![image](https://github.com/ansys/pyaedt/assets/119294121/97eff093-bdb7-4a70-b34e-7c22edcd670a)

-Close and reopen the project to see the icon AddSetup
-Click **AddSetup** icon to execute the script

**Comment by StotreB** [comment]
Hello,

Let's say I have a generic script that I would like to run in my current project, which can be an instance of any EDT app, then how can I load it without having to specify the app? 

Best regards,

**Reply by Samuelopez-ansys** [reply]
@StotreB If you have multiple AEDT sessions, you need to first know the GPRC port of the AEDT session. You can get it from the AEDT Ironpython console: oDesktop.GetGrpcServerPort() or from this pyaedt method you can get all grpc sessions:

from pyaedt.generic.general_methods import grpc_active_sessions

Then you will connect to the AEDT session:

aedt = pyaedt.Desktop(port=50051, version="2024.1")

Once you are connected to the Desktop, you could have multiple projects opened and inside each project you could have multiple designs.

You can get the projects:

aedt.project_list()

The same for the designs of a project:

aedt.design_list(project_name)

And if you have the project name and design name, you can load the aedt app name (HFSS, Icepak...):

aedtapp = aedt[[project_name, design_name]]

I think with all these tools, you can perform any automation.

**Reply by Samuelopez-ansys** [reply]
I missed another way as well.

from pyaedt import get_pyaedt_app

app = pyaedt.Desktop(new_desktop_session=False, specified_version=version, port=port)

active_project = app.odesktop.GetActiveProject()
active_design = active_project.GetActiveDesign()

aedtapp = get_pyaedt_app(design_name=active_design.GetName(), desktop=app)

**Comment by StotreB** [comment]
Thanks a lot! that one seems much easier to implement, thank you
